### ISSUE TITLE
Child condition/where kills WithOne()

### ISSUE THREAD

**Issue by jhovgaard** [ThatRendle/Simple.Data#184]
This one:

``` C#
return Table.Query()
    .Where(Table.Listed == true && Table.PersonId == null)
    .WithOne(Table.Companies.As("Company"))
    .ToList<Job>();
```

Works like a charm! But as soon as I add a condition on the Companies table like this:

``` C#
return Table.Query()
    .Where(Table.Listed == true && Table.PersonId == null)
    .WithOne(Table.Companies.As("Company"))
    .Where(Table.Companies.Plots.CityId == cityId) // <-- This one
    .ToList<Job>();
```

Simple.Data stops casting the Companies table's alias to Company, causing MSSQL to throw this exception on all properties of the `Company`:

```
Simple.Data.Ado.AdoAdapterException: The multi-part identifier "Company.Id" could not be bound.
```

How can I avoid this?

Thanks!

**Comment by ThatRendle** [comment]
Looks like a bug. I'll investigate ASAP.

**Comment by ThatRendle** [comment]
That was a very interesting bug, deserving of some explanation.

At the basic level, the issue here is that in the semantics of your query, Table.Companies.As("Company") and Table.Companies are two different references, and that might be desired behaviour in some circumstances, so your code should work. The reason it didn't is that the join logic saw them as the same reference, whereas the select logic didn't. I have fixed this by changing the equality semantics of the ObjectName class; (amazingly) all the other tests still pass.

With this change, your query will actually join to the Companies table twice, once for the With and once for the Where:

```
SELECT Table.*, Company.* FROM Table
JOIN Companies ON Companies.TableId = Table.Id
JOIN Plots ON Plots.CompanyId = Companies.Id
JOIN Companies Company ON Company.TableId = Table.Id
WHERE Table.Listed = @p1 AND Table.PersonId IS NULL AND Plots.CityId = @p2 
```

(That's not the exact SQL.)

Now, if you don't want that behaviour, what you need to do is specify the aliased table in the Where too:

```
return Table.Query()
    .Where(Table.Listed == true && Table.PersonId == null)
    .WithOne(Table.Companies.As("Company"))
    .Where(Table.Companies.As("Company").Plots.CityId == cityId)
    .ToList<Job>();
```

That's a bit wordy, so I've added an overload for With which works like the one you can use with explicit joins, giving an out parameter which returns the aliased table reference. So you can do this:

```
dynamic company;
return Table.Query()
    .Where(Table.Listed == true && Table.PersonId == null)
    .WithOne(Table.Companies.As("Company"), out company)
    .Where(company.Plots.CityId == cityId)
    .ToList<Job>();
```

These fixes will be in 0.17, which will be released this weekend when I've got a cool new feature working.

**Comment by jhovgaard** [comment]
Great to see you found a solution. It looks great, I didn't even think about specifying alias on my Where().

I love the out solution. Thanks for the fix, looking forward to the weekend :-)
